### PR TITLE
Fix issues with runtime reporting incorrect types for nested generics

### LINKED ISSUE #60565: LLDB crashes because RemoteAST and Reflection metadata diverge
**Describe the bug**
LLDB is crashing with the error message:
```
RemoteAST and runtime diverge $ss18_DictionaryStorageCySO4main8ObserverC12Subscription33_4790D3CC4610A5478AB37A1ABB2801FALLCyAC10LoginEventO_GGD != $ss18_DictionaryStorageCySO4main8ObserverC12Subscription33_4790D3CC4610A5478AB37A1ABB2801FALLCy_AC10LoginEventOGGD
```
The mangled names are related to the program in the reproduction section of this bug report. I've done some preliminary debugging and will provide more information in the "additional context" section.

**Steps To Reproduce**
Small program to reproduce. I called it `Observer.swift` on my machine.
```
import Foundation

public enum LoginEvent {
    case start
    case success
    case error(error: String)
}

public class Observer<Event> {
    public init() {}

    @discardableResult public func subscribe() -> ObjectIdentifier {
        let subscription = Subscription()
        let identifier = ObjectIdentifier(subscription)
        self.subscriptions[identifier] = subscription
        return identifier
    }

    private class Subscription {}

    private var subscriptions: [ObjectIdentifier: Subscription] = [:]
}


func main() {
    let foo = Observer<LoginEvent>()
    foo.subscribe()
    print("Hello world!") // break here
}

main()
```
Compile with swiftc like so:
```
swiftc -g -sdk $(xcrun --show-sdk-path) Observer.swift -o Observer
```

Launch lldb to debug the program like so:
```
lldb -- Observer
```
Place a breakpoint on line 28 (The line with print in `main()`) and run until you hit that breakpoint. Try to get the contents of `foo` with `v foo` and observe the crash.

**Expected behavior**
I expect LLDB to not crash. Additionally, I expect LLDB to give me a summary of the contents of the variable `foo` in the example program. 

**Environment (please fill out the following information)**
 - OS: macOS Monterey 12.5
 - Chip: Apple M1 Max
 - Xcode Version/Branch: Reproduces on the swift "main" branch as well as the 5.7 release branch. I built against Xcode 13.4.

**Additional context**
When I had debugged the problem, it looks like the type reconstructed from reflection metadata is actually incorrect. I debugged LLDB and the type that was reconstructed from reflection metadata looks like this:
```
(bound_generic_class Swift._DictionaryStorage
  (struct Swift.ObjectIdentifier)
  (bound_generic_class Observer.Observer.(Subscription in _FA454F2F3490BDD2D6AE0E5A6DDA2566)
    (enum LoginEvent.LoginEvent)))
```
This suggests that `Observer.Observer.Subscription` is generic (i.e. that the type looks like `Observer.Observer.Subscription<LoginEvent>`). This is definitely incorrect as the real type looks like `Observer.Observer<LoginEvent>.Subscription`. The type I received from the AST looks far more correct:
```
(bound_generic_class_type decl=Swift.(file)._DictionaryStorage
  (struct_type decl=Swift.(file).ObjectIdentifier)
  (class_type decl=Observer.(file).Observer.Subscription
    (parent=bound_generic_class_type decl=Observer.(file).Observer
      (enum_type decl=LoginEvent.(file).LoginEvent))))
```
Curiously, I'm not sure if the reflection metadata is necessarily incorrect. Using `swift-reflection-dump` to dump out the information in the binary, the type of `subscriptions` is reported correctly:
```
subscriptions: Swift.Dictionary<Swift.ObjectIdentifier, Observer.Observer<A>.(Subscription in _FA454F2F3490BDD2D6AE0E5A6DDA2566)>
(bound_generic_struct Swift.Dictionary
  (struct Swift.ObjectIdentifier)
  (class Observer.Observer.(Subscription in _FA454F2F3490BDD2D6AE0E5A6DDA2566)
    (bound_generic_class Observer.Observer
      (generic_type_parameter depth=0 index=0))))
```

### PR DESCRIPTION
**Pull request by NuriAmari**
Resolves #60565

### CODE REVIEW

**Comment by NuriAmari** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
Can you add a test case?

**Comment by NuriAmari** [comment]
@swift-ci Please smoke test

**Comment by NuriAmari** [comment]
Tests created in this PR: https://github.com/apple/llvm-project/pull/5544. I've never contributed to Apple's fork of LLVM -- how do I run the LLDB tests against this PR (they will fail otherwise)? Also am I merging into the correct branch [apple:stable/20220421](https://github.com/apple/llvm-project/tree/stable/20220421)?

Many thanks to @augusto2112 for extensively helping me find the fix.

While this patch seems to fix the issue, I am still fuzzy on the how everything works -- so please let me know if there is a more correct approach to take here.

**Comment by augusto2112** [comment]
@NuriAmari thank you for this!

> how do I run the LLDB tests against this PR (they will fail otherwise)? 

From the lldb side you can type the following:

```
https://github.com/apple/swift/pull/61819
"at" swiftci test (with an @, don't want to trigger it here)
```

> Also am I merging into the correct branch [apple:stable/20220421](https://github.com/apple/llvm-project/tree/stable/20220421)?

Yes, that's correct. If you could also cherrypick it into `origin/stable/20221013` and `next` after merging, that'd be helpful.

> While this patch seems to fix the issue, I am still fuzzy on the how everything works -- so please let me know if there is a more correct approach to take here.

So, context descriptors have a flat representation of generic parameters, while typerefs use a nested representation. Lets say you have this type:
```
class A<T> {
  class B<U, V> {
    class C<W> {
    }
  }
}
```
In the context descriptor representation, `C` has 4 generic parameters, `B` has 3 and `A` has 1. While in the typeref representation, `A` and `C` have 1 generic parameter and `B` has 2.

What this lambda is doing is going up the context descriptor chain and building a vector with the shape that the typeref hierarchy expects. 

Does that make sense?

File: include/swift/Remote/MetadataReader.h [hunk 1]
```diff
@@ -2943,6 +2946,9 @@ class MetadataReader {
               if (auto parentContext = parentContextRef->getResolved())
                 countLevels(parentContext, runningCount);
 
+          if (!dyn_cast<TargetTypeContextDescriptor<Runtime>>(current))
```
[the remark this answers is not in the document]
Comment (augusto2112) [reply include/swift/Remote/MetadataReader.h:2949]: The only worry I have here is what happens if you have a function involved. You can't have types defined generic function, but you could have something like:
```
func a() 
  class B<T>{
    func f() {
      print(1) // break here, v self
    }
  }
  B<Int>().f()
}
a()
```
Could you test what happens in this case?

**Review by augusto2112 — CHANGES_REQUESTED** [review]

**Comment by NuriAmari** [comment]
> If you could also cherrypick it into origin/stable/20221013 and next after merging, that'd be helpful.

Ok will do thanks!

> Does that make sense?

Yes the part that remains a little unclear to me is exactly which contexts we need to consider when walking up the chain. I know when we discussed skipping anonymous contexts, you mentioned they are introduced by private type declarations. I will try again but I couldn't verify this is a small experiment. Overall what kind of contexts there are, and which we care about is what I would like to clarify.

**Comment by augusto2112** [comment]
> Overall what kind of contexts there are, and which we care about is what I would like to clarify.

You can check in `MetadataValues.h`, you have:
```
enum class ContextDescriptorKind : uint8_t {
  /// This context descriptor represents a module.
  Module = 0,
  
  /// This context descriptor represents an extension.
  Extension = 1,
  
  /// This context descriptor represents an anonymous possibly-generic context
  /// such as a function body.
  Anonymous = 2,

  /// This context descriptor represents a protocol context.
  Protocol = 3,
  
  /// This context descriptor represents an opaque type alias.
  OpaqueType = 4,

  /// First kind that represents a type of any sort.
  Type_First = 16,
  
  /// This context descriptor represents a class.
  Class = Type_First,
  
  /// This context descriptor represents a struct.
  Struct = Type_First + 1,
  
  /// This context descriptor represents an enum.
  Enum = Type_First + 2,
  
  /// Last kind that represents a type of any sort.
  Type_Last = 31,
};
```

You can add the following code in the lambda to check what type the context descriptor represents:
```
          auto node = buildContextMangling(current, dem);
          node->dump();
```

For something like:
```
class A<T> {
  private class B<U> {
    func f<C>(t : C) {
      print(1) // br here
    }
  }
  func g() {
    B<Double>().f(t: true)
  }
}
let a = A<Int>()
a.g()
```

This will print:
```
// Module descriptor
kind=Module, text="t"

//  Class descriptor for `A`
kind=Type
  kind=Class
    kind=Module, text="t"
    kind=Identifier, text="A"

// Anonymous descriptor, it "nests" the class B
kind=AnonymousContext
  kind=Identifier, text="$100003ce0"
  kind=Class
    kind=Module, text="t"
    kind=Identifier, text="A"

// Class descriptor for B
kind=Type
  kind=Class
    kind=Class
      kind=Module, text="t"
      kind=Identifier, text="A"
    kind=PrivateDeclName
      kind=Identifier, text="_9AA20C7C53A214BFCE5DF8269DEEAC4F"
      kind=Identifier, text="B"
```

File: include/swift/Remote/MetadataReader.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (NuriAmari) [reply include/swift/Remote/MetadataReader.h:2949]: Seems it fails an assert and crashes as a result:
```
(lldb) target create "Functions"
Current executable set to '/Users/nuriamari/git/repro-pad/T129460057-[pika]-LLDB-crashes-due-to-difference-between-RemoteAST-and-Reflection-metadata/Functions' (arm64).
(lldb) b Functions.swift:4
Breakpoint 1: where = Functions`B.f() -> () in Functions.a() -> () + 60 at Functions.swift:4:13, address = 0x00000001000058e4
(lldb) r
Process 74769 launched: '/Users/nuriamari/git/repro-pad/T129460057-[pika]-LLDB-crashes-due-to-difference-between-RemoteAST-and-Reflection-metadata/Functions' (arm64)
/Users/nuriamari/git/swift-project/swift/lib/Demangling/Remangler.cpp:3685: assertion failed for Node 0x15a5b3910: unboundType->getKind() == Node::Kind::Type
/Users/nuriamari/git/swift-project/swift/lib/Demangling/Remangler.cpp:3685: Node 0x15a5b3910 is:
kind=Function
  kind=Module, text="Functions"
  kind=Identifier, text="a"
  kind=Type
    kind=FunctionType
      kind=ArgumentTuple
        kind=Type
          kind=Tuple
      kind=ReturnType
        kind=Type
          kind=Tuple
```
Still looking into it.

**Comment by NuriAmari** [comment]
@swift-ci Please smoke test

File: include/swift/Remote/MetadataReader.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (NuriAmari) [reply include/swift/Remote/MetadataReader.h]: Works after the latest changes

File: include/swift/Reflection/TypeRefBuilder.h [hunk 2]
```diff
@@ -572,9 +570,19 @@ class TypeRefBuilder {
         args.end() - argsIndex - numGenericArgs, args.end() - argsIndex);
 
     const BoundGenericTypeRef *parent = nullptr;
-    if (node->hasChildren())
-     parent = createBoundGenericTypeReconstructingParent(
-        node->getFirstChild(), decl, --shapeIndex, args, argsIndex + numGenericArgs);
+    if (node->hasChildren()) {
+      // Skip over nodes that are not of type class, enum or struct
+      auto parentNode = node->getFirstChild();
+      while (parentNode->getKind() != Node::Kind::Class &&
```
Comment (augusto2112) [inline include/swift/Reflection/TypeRefBuilder.h:576]: There could be a null pointer deref here.

**Review by augusto2112 — APPROVED** [review]
LGTM with comment fixed!

File: include/swift/Reflection/TypeRefBuilder.h (hunk 2 above)
Comment (NuriAmari) [reply include/swift/Reflection/TypeRefBuilder.h:576]: Let me know if this looks reasonable

**Comment by NuriAmari** [comment]
@swift-ci Please smoke test

File: include/swift/Reflection/TypeRefBuilder.h (hunk 2 above)
Comment (augusto2112) [reply include/swift/Reflection/TypeRefBuilder.h:576]: LGTM :)

File: include/swift/Remote/MetadataReader.h [hunk 3]
```diff
@@ -2906,9 +2906,11 @@ class MetadataReader {
   template <
       typename T = BuilderType,
       typename std::enable_if_t<
-          !std::is_same<
-              bool,
-              decltype(T::needsToPrecomputeParentGenericContextShapes)>::value,
+          !(std::is_same<
+                const bool,
+                decltype(T::needsToPrecomputeParentGenericContextShapes)>::
+                value &&
```
Comment (kastiglione) [inline include/swift/Remote/MetadataReader.h:2912]: did clang-format put the `value` on a separate line from its `::`?

Comment (NuriAmari) [reply include/swift/Remote/MetadataReader.h:2912]: Yes:
```
~/git/swift-project/swift (remote-ast-runtime-type-mismatch) git-clang-format HEAD~1
changed files:
    include/swift/Remote/MetadataReader.h
~/git/swift-project/swift (remote-ast-runtime-type-mismatch) git diff
diff --git a/include/swift/Remote/MetadataReader.h b/include/swift/Remote/MetadataReader.h
index 376135f3afe..c1faf54f89d 100644
--- a/include/swift/Remote/MetadataReader.h
+++ b/include/swift/Remote/MetadataReader.h
@@ -2908,8 +2908,9 @@ private:
       typename std::enable_if_t<
           !(std::is_same<
                 const bool,
-                decltype(T::needsToPrecomputeParentGenericContextShapes)>::value
-                && T::needsToPrecomputeParentGenericContextShapes),
+                decltype(T::needsToPrecomputeParentGenericContextShapes)>::
+                value &&
+            T::needsToPrecomputeParentGenericContextShapes),
           bool> = true>
   BuiltTypeDecl buildNominalTypeDecl(ContextDescriptorRef descriptor) {
     // Build the demangling tree from the context tree.
```

Comment (kastiglione) [reply include/swift/Remote/MetadataReader.h:2912]: that's unfortunate!

File: include/swift/Reflection/TypeRefBuilder.h [hunk 4]
```diff
@@ -572,9 +570,23 @@ class TypeRefBuilder {
...
+             parentNode->getKind() != Node::Kind::Structure &&
+             parentNode->getKind() != Node::Kind::Enum) {
+        if (parentNode->hasChildren()) {
+          parentNode = parentNode->getFirstChild();
```
Comment (kastiglione) [inline include/swift/Reflection/TypeRefBuilder.h:580]: Two questions:

1. Is it always the case that the nodes of interest will be the first child?
2. The recursion is indiscriminate about which nodes it passes over, are there any nodes it should stop recursing on?

File: include/swift/Reflection/TypeRefBuilder.h [hunk 5]
```diff
@@ -558,9 +558,7 @@ class TypeRefBuilder {
     auto kind = node->getKind();
     // Kinds who have a "BoundGeneric..." variant.
     if (kind != Node::Kind::Class && kind != Node::Kind::Structure &&
-        kind != Node::Kind::Enum && kind != Node::Kind::Protocol &&
-        kind != Node::Kind::OtherNominalType && kind != Node::Kind::TypeAlias &&
-        kind != Node::Kind::Function)
```
Comment (kastiglione) [inline include/swift/Reflection/TypeRefBuilder.h:563]: @augusto2112 you added these non-{class,struct,enum} types originally, were they needed originally but not now?

Comment (augusto2112) [reply include/swift/Reflection/TypeRefBuilder.h:563]: No, that was a bug. We found that the demangler only deals with classes structs and enums when building a detangle tree from a typeref https://github.com/apple/swift/blob/main/stdlib/public/runtime/Demangle.cpp#L153-L157

File: include/swift/Reflection/TypeRefBuilder.h (hunk 4 above)
Comment (augusto2112) [reply include/swift/Reflection/TypeRefBuilder.h:580]: 1. My understanding is that yes, it'll always be the first child. 

2. Yes. When I wrote the old code I assumed that any kind that had a generic counterpart was valid. As per my other comment, we found that the only ones that are handled (for generic typerefs) are structs classes and enums.

**Pull request merged**